Improve data workflow documentation

ekatef opened the pull request:
## Changes proposed in this Pull Request

Add some points to the data workflow documentation. The aim is to facilitate model building for an arbitrary world region.

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

- [x] Check pictures scale and alignment
- [x] Add a list of actions needed to do the validation

davide-f commented:
@ekatef thanks for the PR and pointing out some lacking parts.
I think the structure looks good, including also the locations of the new rst files into the documentation.

As a little suggestion, the "data_workflow" may be placed under the introduction rst as a new block.
Given its content, it may be interesting to rename it into something like what data are used, because "Data workflow" may be misleading.
Parts of the content of that section may be taken and updated from: https://pypsa-meets-earth.readthedocs.io/en/latest/introduction.html#licence
In that area you can find some references to websites and data license that are worth mentioning in that section

ekatef commented:
> @ekatef thanks for the PR and pointing out some lacking parts. I think the structure looks good, including also the locations of the new rst files into the documentation.
> 
> As a little suggestion, the "data_workflow" may be placed under the introduction rst as a new block. Given its content, it may be interesting to rename it into something like what data are used, because "Data workflow" may be misleading. Parts of the content of that section may be taken and updated from: https://pypsa-meets-earth.readthedocs.io/en/latest/introduction.html#licence In that area you can find some references to websites and data license that are worth mentioning in that section

@davide-f, thank you so much for your suggestions and the discussion. For me it looks like we are moving closer to the model accessibility on the global scale which feels very good :)

Have tried to implement some improvements:
- moved `data_workflow` under the `introduction`;
- tried to improve `data_workflow` content adding a general introduction;
- attached validation suggestions;
- moved data licensing info into `data_workflow`.

Regarding licenses not sure if they are now good placed into the `data_workflow`. My personal preference is to have all the licenses information collected in one place rather than disperse. Although, I do agree that some information regarding data sources is needed in the `data_workflow` section. Would it probably make sense to replace licensing information in `data_workflow` with references/acknowledgements?

pz-max commented:
Hey @ekatef , jsut heads up. I changed recently the docs: https://github.com/pypsa-meets-earth/pypsa-earth/pull/476, https://github.com/pypsa-meets-earth/pypsa-earth/pull/474 . I think most of your changes are independent. Looking forward to your improvements :)

ekatef commented:
Hey @pz-max, thank you for pointing out! I have still noticed that the docs were updated and that's great :) I'll try to work out a minimally readable addition to the docs as soon as I can.

ekatef commented:
Hey @AnasAlgarei! This PR is aimed to help in changing the modeling domain (generating cutout etc.). Probably, some of the points may be helpful for you in increasing the model coverage. Will be happy to elaborate/explain any of them if you feel it'd be helpful. And will be very grateful for any comments and suggestions :)

AnasAlgarei commented:
> Hey @AnasAlgarei! This PR is aimed to help in changing the modeling domain (generating cutout etc.). Probably, some of the points may be helpful for you in increasing the model coverage. Will be happy to elaborate/explain any of them if you feel it'd be helpful. And will be very grateful for any comments and suggestions :)

Hey @ekatef! Excellent modifications, it will help a lot for people to get started and working in a larger scale, definitely will be of help to me too, I will try and make some suggestions to improve it as we go with the EDF project, thank you :)

ekatef commented:
@davide-f, I think it's ready for review. 

My feeling is that it's better to leave the data licences in the introduction along with the code licences. I have done an attempt to split code and data licensing parts but the result was not really persuading :)

A quick additional question: what was the data source for hydro_capacities.csv? I think it's worth documenting along with other datasets.

davide-f commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,102 @@
+Here we'll look into architecture of the data workflow while practical hand-ons are given in the Tutorial section.
+
+**1. Grid topology data**
+OpenStreetMap `OSM <https://www.openstreetmap.org/>`_ data are used to build power grid topology model. OSM is the biggest crowd-sourced collection of geographic information, which is daily updated and includes geolocation references. The OSM data are being loaded by `download_osm_data` and cleaned by `clean_osm_data` rules, respectively.
```
Sample images are already available in the img folder of the doc, mainly related to the europe case, but we can start from there.
To add the command .. image should be used. You can use examples from other rst files.
E.g. 
.. image:: img/Flowchart_base.png
    :width: 80%

ekatef replied:
Agree that is's always good to have some good graphics. Have added the pictures which we still have as the first iteration. Although, their placement is likely to be non-optimal right now. I'll have to play around Sphinx a bit to make the output more visually appealing

ekatef replied:
Done. Obviously there is room for improvement but now it still looks decent.

davide-f commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,102 @@
+
+Currently we are using the following resources.
+
+- *environmental*
```
The *environmental* may be formatted using titles in rst format. See examples here: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#sections

ekatef replied:
Done!

davide-f commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,102 @@
+
+**landcover** describes the shapes of world protected areas that are needed to identify in what areas no (renewable) assets can be installed. Currently are used to generate a `natura.tiff` raster. Will be deprecated once the global `natura.tiff` will be available.
+
+**osm** are raw `OpenStreetMap <https://www.openstreetmap.org/>`_ data. They are being loaded when running the `download_osm_data` rule.
```
Isn't this osm redundant with respect to above?

ekatef replied:
Thanks a lot! Fixed.

davide-f commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,102 @@
+
+.. .. _data_validation_tips:
+
+.. Data validation tips
```
What about this part?

ekatef replied:
Removed

davide-f commented on doc/how_to_customize.rst:
```diff
@@ -0,0 +1,54 @@
+.. _how_to_customize:
+
+##########################################
+How to customize?
```
I'd say: how to use PyPSA-Earth to your energy problem? Or something like that.
Customization may be interpreted how to modify also the source for example. Or maybe this is what I may also understand (I may be wrong though)

ekatef replied:
Agree, that is not obvious that data adjustment is meant. Tried to improve.

davide-f commented on doc/how_to_customize.rst:
```diff
@@ -0,0 +1,54 @@
+2) the countries of interest defined with `countries` list in the `config.yaml` should be covered by the cutout area;
+3) the cutout time dimension, the weather year used for demand modeling and the actual snapshot should match.
+
+**2. Load all the common data**
```
Maybe "Customize the input data toiyour national-specific data"?

ekatef replied:
Changed

davide-f commented on doc/how_to_customize.rst:
```diff
@@ -0,0 +1,54 @@
+
+Mostly these data are global and should be loaded only once when installing the model. The exception is the cutout which requires of about ~200 Gb in it's global version. The cutout which is supplied with the model could be loaded by running `retrieve_databundle_light` rule, corresponds to Africa. If you are interested in other part of the world a custom cutout should be built. 
+
+**3. Build the custom cutout**
```
I love it! :)

ekatef replied:
Very happy to hear :)

davide-f commented on doc/how_to_customize.rst:
```diff
@@ -0,0 +1,54 @@
+
+The `build_cutout` flag should be set `true` to generate the cutout. After the cutout is ready, it's recommended to set `build_cutout` to `false` to avoid overwriting the existing cutout by accident.
+
+**4. Build a natura.tiff raster**
```
Ok, though the idea is to provide a general default global value (sooner or later).
That is worth mentioning

ekatef replied:
Amended a comment to the previous (point 2) part

davide-f commented on doc/how_to_start.rst:
```diff
@@ -0,0 +1,22 @@
+##########################################
+How to start?
+##########################################
+
```
This seems to be empty

ekatef replied:
Let's remove it for now. I think it's worth to supply a very basic starter kit to demonstrate what the model is capable of. But it seems to be a separate task

davide-f commented on doc/how_to_validate.rst:
```diff
@@ -0,0 +1,35 @@
+How to validate?
+##########################################
+
+Data workflow is PyPSA-Earth allows to extract all the data needed to build your energy model from global open data sources. However, data availability for many parts of the world is still quite limited and it's essential to validate the data included into the model. Usually the best sources to compare with are regional data hubs along with harmonized datasets of international organisations such  ...
```
I'd start highlighting the sample notebook we have.
Then, I'd add a list of actions needed to do the validation. The list can be derived by the notebook itself

ekatef replied:
Agree and will elaborate that

ekatef replied:
Done

davide-f commented on doc/index.rst:
```diff
@@ -111,16 +111,24 @@ Documentation
 
    introduction
+   data_workflow
    installation
```
May it be better to locate the data_workflow after the installation?

ekatef replied:
Done :)

davide-f commented:
Regarding the data license, absolutely, they shall stay there.
My prior comment was to just take the links and list of inputs to have the panorama of what data are used.

Regarding the hydro_capacity, let's discuss it today

ekatef commented:
Thank you so much for the review @davide-f! That is an absolute pleasure to work with you :)

According to my feeling, this PR almost done. I have only to check if the embedded pictures look properly and improve a bit validation instructions.

davide-f commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,100 @@
+There are a number datasets applied in PyPSA-Earth to build a realistic model. Original datasets are stored in the `pypsa-earth/data/` folder. Currently we are using the following resources.
+
+Environmental
+------------------------------------
```
If we use the "-----" for the 1. 2. 3. titles, then there would be the need to revise this title.
In this case, we could use also nested lists something like:
- list 1
- list 2
  - 2.1
  - 2.2
 
for example

davide-f commented on doc/how_to_fit.rst:
```diff
@@ -0,0 +1,68 @@
+
+3. the cutout time dimension, the weather year used for demand modeling and the actual snapshot should match.
+
+2. Load all the common data
```
This may be confusing, I'd remove this part as the workflow takes care of this.
It seems that to run the model there is the need to load all the common data first, which is not the case

ekatef replied:
Agree. How to remove the common data should be clear from the installation instructions

davide-f commented on doc/how_to_fit.rst:
```diff
@@ -0,0 +1,68 @@
+Mostly these data are global and should be loaded only once when installing the model. The exception is the cutout which requires of about ~200 Gb in it's global version. The cutout which is supplied with the model could be loaded by running `retrieve_databundle_light` rule, corresponds to Africa. If you are interested in other part of the world a custom cutout should be built. In the future we p ...
+
+3. Build the custom cutout
+=========================================
```
Right after this title I'd add a note, like a discaimer

.. note::
   Skip this recommendation if the region of your interest is within Africa and you are fine with the 2013 weather year

I think the user shall know immediately when this task can be skipped

ekatef replied:
Done

davide-f commented on doc/how_to_fit.rst:
```diff
@@ -0,0 +1,68 @@
+The `build_cutout` flag should be set `true` to generate the cutout. After the cutout is ready, it's recommended to set `build_cutout` to `false` to avoid overwriting the existing cutout by accident.
+
+4. Build a natura.tiff raster
+=========================================
```
Similarly to point 3, I'd add a note to highlight when this feature can be skipped

ekatef replied:
Done. According to my impression, both notes have made the overall section  much nicer visually :)

davide-f commented on doc/how_to_fit.rst:
```diff
@@ -0,0 +1,68 @@
+How to use PyPSA-Earth for your energy problem?
+##########################################
+
+PyPSA-Earth can be tailored to represent any part of the world quite easily. The following procedure may be recommended.
```
may be -> are

ekatef replied:
Changed

davide-f commented on doc/how_to_validate.rst:
```diff
@@ -0,0 +1,64 @@
+
+.. TODO add a list of actions needed to do the validation
+
+Data workflow is PyPSA-Earth allows to extract all the data needed to build your energy model from global open data sources. An example of the validation procedure is available in the `Nigeria validation <https://github.com/pypsa-meets-earth/documentation/blob/main/notebooks/validation/validation_nigeria.ipynb>`_ notebook. Public information on the power system of Nigeria are compared to those ob ...
```
I'd remove the first sentence and be direct on what this section is about.
Like "To validate the data obtained with PyPSA-Earth, we recommend to go through the procedure here detailed. An exampled of ..."

ekatef replied:
Revised

davide-f commented on doc/how_to_validate.rst:
```diff
@@ -0,0 +1,64 @@
+
+    #. energy mix.
+
+Data sources
```
"Third-party data sources for validation" or something like that? feel free to improve on it.
The message below should be: to do the validation you need to compare the data obtained with the workflow with other (third-party) trustworthy data, such as those that follow

ekatef replied:
Tried to improve

davide-f reviewed (commented):
Another round :)

ekatef commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,100 @@
+Data used by the model
+##########################################
+
+PyPSA-Earth has the flexible scope when a modeling domain can be selected from the whole Earth to any subregion. The modeling workflow includes data management workflow which is focused on open data collection, modification, prediction and validation. The developed data processing approaches aimed to provide model-ready data with temporal and spatial resolution as high as it's needed by a specifi ...
```
Done!

ekatef commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,100 @@
+
+Here we'll look into architecture of the data workflow while practical hand-ons are given in the Tutorial section.
+
+1. Grid topology data
```
I think we may safely delete "Data management strategy" with some benefits for overall clarity :)

ekatef commented:
> Another round :) Almost there :D

Thank you so much :-) 
Have added a quick fix for retrieval rules.

ekatef commented:
I think this documentation addition may be helpful to add some hand-ons to #505. Not sure however what is the best way to incorporate `[how_to_fit](https://github.com/pypsa-meets-earth/pypsa-earth/pull/467/files#diff-187752b18d7d9918f1f039d87b10c699d34f3b43e2cc095f01c0e553bd3aad56)` and `[how_to_validate](https://github.com/pypsa-meets-earth/pypsa-earth/pull/467/files#diff-c96c09823cf41ec761e386890304d81918c1203f1c57154c383acea737cb8f08)` into the existing rst-doc structure: it looks quite logical to add both sections to the `Tutorial` section. But wouldn't it make the Tutorial too long?

@AnasAlgarei, @pz-max, @davide-f, any thoughts?

ekatef commented:
@pz-max, I think it's ready for your review.

pz-max commented:
@ekatef added some fixes and clarifications. Are the suggestions ok with you?
Happy to merge afterwards. Thanks for the fantastic docs additions :)

davide-f commented on doc/index.rst:
```diff
@@ -110,7 +110,9 @@ Documentation
+* :doc:`short_tutorial`
 * :doc:`tutorial`
+* :doc:`data_workflow`
 * :doc:`notebooks`
```
It should be a good idea to change the name of the notebooks section, or maybe integrate the notebooks section into the file tutorial.rst

Maybe nice to revise the name as "notebooks" is a technical word, they may be explanatory examples but I'm fine with the current name for the time being

ekatef replied:
Changed to "Examples"

ekatef commented on doc/data_workflow.rst:
```diff
@@ -0,0 +1,95 @@
+    :height: 150 px
+    :align: center
+
+The raw OSM data are being loaded when running the `download_osm_data` rule and stored in the folder `pypsa-earth/data/osm/{region}/pbf/`. Here `region` denotes a continent, e.g. Africa, or a macro region, e.g. Central America, where the countries of interest belong. The pbf-files contain the entire OSM data for the country; the specific network information related to generators, substations, lin ...
```
Done

ekatef commented on doc/short_tutorial.rst:
```diff
@@ -0,0 +1,101 @@
+
+
+In the configuration file `config.yaml` there is a flag `retrieve_databundle` which triggers data loading and a `tutorial` flag which
+determines that the loaded data belong to the light tutorial kit:
```
Fixed for "light" and a countries list. Regarding the computational requirements, we still give some details on that when [introducing](https://github.com/ekatef/pypsa-earth/blob/data_workflow_docs/doc/short_tutorial.rst#how-to-build-the-tutorial-model) the tutorial

ekatef commented on doc/tutorial.rst:
```diff
@@ -10,65 +10,8 @@
--------------------------------------------------
+======================================================
 
 The model can be adapted to include any country, even multiple countries (e.g. Nigeria and Benin) or continents
```
Have added a clarification

ekatef commented on doc/short_tutorial.rst:
```diff
@@ -0,0 +1,101 @@
+
+
+##########################################
+Short Tutorial
```
Tried to improve :)

ekatef commented on doc/tutorial.rst:
```diff
@@ -10,65 +10,8 @@
 Tutorial
 ##########################################
```
Mmm... Agree with Max that it's worth to keep some similarities with PyPSA-Earth/PyPSA docs. Have kept "Tutorial" as far but don't mind modify it further

pz-max replied:
yeah let's keep it aligned with PyPSA-Eur as much as possible.
Short tutorial & Tutorial aligns well with Snakemake.
Anyways, I think these are details where everyone has different preferences :)

davide-f replied:
Yeah, I was looking at the standard of pypsa that I liked ("Quick Start") https://pypsa.readthedocs.io/en/latest/

But its a detail anyway

ekatef commented:
@pz-max, thanks a lot for the introduced improvements. I think they look absolutely great.

@davide-f, many thanks  for the review. Added a modified picture a the side repository created specifically for this purpose. Would be happy to re-organise in a more proper way but not sure how to do it better. Have you suggested to commit a picture into `documentation/doc/`?

Generally, it looks like we have one more demonstration that naming is one of two really hard things :D Have changed the sections titles (but kept files names untouched at the moment to avoid unnecessary renaming). Will be happy to make adjustments

davide-f commented:
@ekatef awesome! :)
The content of the folder "documentation/doc/img/*" is copied into "pypsa-earth/doc/img/*"
So in the documentation, you can expect the same tree structure of the files into the doc/img of documentation to be copied into the other folder.
It may be worth adding a very simple README into pypsa-earth/doc/img to highlight that... feel free to do it

Hence, feel free to drop the images into documentation/doc/img

ekatef commented:
@davide-f, thanks for clarifications!  It worked with with "documentation/doc/img/". I have added a [PR with the updated picture](https://github.com/pypsa-meets-earth/documentation/pull/19) to `documentation`.

Have amended a short instruction how to add an image both as README to pypsa-earth/doc/ (pypsa-earth/doc/img is not tracked) and into `how-to-doc` section in `read-the-docs`

pz-max commented:
@ekatef thanks for the great PR. Merged this now. 
Any additional fixes can be addressed in new PR's.
